heat/core/linalg/solver.py

```diff
@@ -106,15 +106,23 @@ def lanczos(
     n, column = A.shape
     if n != column:
         raise TypeError("Input Matrix A needs to be symmetric.")
-    T = ht.zeros((m, m))
+    T = ht.zeros((m, m), dtype=A.dtype, device=A.device)
     if A.split == 0:
         # This is done for better memory access in the reorthogonalization Gram-Schmidt algorithm
         V = ht.ones((n, m), split=0, dtype=A.dtype, device=A.device)
     else:
         V = ht.ones((n, m), split=None, dtype=A.dtype, device=A.device)
 
     if v0 is None:
-        vr = ht.random.rand(n, split=V.split)
+        try:
+            vr = ht.random.rand(n, split=V.split, dtype=V.dtype, device=V.device)
+        except ValueError:
+            # V is complex
+            vr_dtype = ht.float64 if V.dtype is ht.complex128 else ht.float32
+            vr = (
+                ht.random.rand(n, split=V.split, dtype=vr_dtype, device=V.device)
+                + ht.random.rand(n, split=V.split, dtype=vr_dtype, device=V.device) * 1j
+            )
```

Comment:
It's ok for me, but there a two points that come to my mind:
- `ValueError` and then inferring that it must be complex might be a heuristic that currently works, but maybe not in the future. One could alternatively also add a check at function start that the supported dtypes are supplied?
- Considering the scope and regime in which lanczos works, I am wondering whether supporting complex numbers sends the right message to the user. Maybe it is not too bad to flag a ValueError in the complex case to highlight that s.p.d. matrices (and hence real-valued `v0`) should be supplied. One can even tell the user that if he thinks the matrix is s.p.d. that `A.real` would already do the trick. On the plus side, this would save half of the unnecessary computations.

Reply:
Agreed on the heuristic, updated. I've also updated the docs.